pytket/extensions/cutensornet/tnstate/ttn.py

```diff
@@ -0,0 +1,845 @@
+                is smaller than one.
+        """
+        self._lib = libhandle
+        self._cfg = config
+        self._logger = set_logger("TTN", level=config.loglevel)
+        self.fidelity = 1.0
+        self.nodes: dict[RootPath, TreeNode] = dict()
+        self.qubit_position: dict[Qubit, tuple[RootPath, int]] = dict()
+
+        if self._cfg.truncation_fidelity < 1:
+            raise NotImplementedError(
+                "Truncation fidelity mode not currently implemented on TTN."
+            )
+
+        n_groups = len(qubit_partition)
+        if n_groups == 0:  # There's no initialisation to be done
+            return None
+        if n_groups == 1:
+            raise ValueError(
+                "Only one entry to qubit_partition provided."
+                "Introduce a finer partition of qubits."
+            )
+
+        n_levels = math.floor(math.log2(n_groups))
+        if n_groups != 2**n_levels:
+            raise ValueError(
+                "The number of entries in qubit_partition must be a power of two."
+            )
+
+        # Create the TreeNodes of the different groups of qubits
+        for k, qubits in qubit_partition.items():
+            if k < 0 or k >= n_groups:
+                raise ValueError(
+                    f"The keys of qubit_partition must range from 0 to {n_groups-1}."
+                )
+
+            # Calculate the root path of this group
+            path = []
+            for l in reversed(range(n_levels)):
+                if k < 2**l:
+                    path.append(DirTTN.LEFT)
+                else:
+                    path.append(DirTTN.RIGHT)
+                    k -= 2**l
+
+            # Add each qubit to the qubit_position dictionary
+            for i, q in enumerate(qubits):
+                if q in self.qubit_position:
+                    raise ValueError(
+                        f"Qubit {q} appears in multiple entries of qubit_partition."
+                    )
+                self.qubit_position[q] = (tuple(path), i)
+
+            # This tensor has a physical bond per qubit and one virtual bond at the
+            # end for the parent (dim=1)
+            shape = tuple([2] * len(qubits) + [1])
+            # Initialise the tensor of this group of qubits to |0>
+            tensor = cp.zeros(shape=shape, dtype=self._cfg._complex_t)
+            ket_zero_entry = tuple(0 for _ in shape)  # Index 0 on all bonds
+            tensor[ket_zero_entry] = 1  # Amplitude of |0> set to 1
+
+            # Create the TreeNode
+            node = TreeNode(tensor, is_leaf=True)
+            self.nodes[tuple(path)] = node
+
+        # Create the internal TreeNodes
+        paths: list[list[DirTTN]] = [[]]
+        for _ in range(n_levels):
+            # Create the TreeNode at this path
+            for p in paths:
+                tensor = cp.ones(shape=(1, 1, 1), dtype=self._cfg._complex_t)
+                self.nodes[tuple(p)] = TreeNode(tensor)

```

Comment:
Not sure how I feel for the root node's key being an empty tuple... But it works, so probably no biggie!

Reply:
It's the empty path, what makes you uneasy about it?

Reply:
I am not really sure, never used an empty struct as a key :) But yeah, should be fine.